File: shield-core/src/fingerprint.rs
//! Hardware fingerprinting for device-bound encryption.
//!
//! Collects platform-specific hardware identifiers to create device-bound keys.
//! Adapted from SaaSClient-SideLicensingSystem with enhanced cross-platform support.

use crate::error::{Result, ShieldError};
use std::process::Command;

/// Fingerprint collection mode.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum FingerprintMode {
    /// No hardware fingerprinting (backward compatible).
    None,
    /// Use motherboard serial number only.
    Motherboard,
    /// Use CPU identifier only.
    CPU,
    /// Use combined motherboard + CPU (recommended).
    Combined,
}

impl Default for FingerprintMode {
    fn default() -> Self {
        Self::None
    }
}

/// Collect hardware fingerprint based on mode.
///
/// # Errors
/// Returns error if hardware identifiers cannot be collected.
///
/// # Platform Support
/// - **Windows**: wmic commands (baseboard, CPU)
/// - **Linux**: /sys/class/dmi/id, dmidecode, /proc/cpuinfo
/// - **macOS**: system_profiler SPHardwareDataType
pub fn collect_fingerprint(mode: FingerprintMode) -> Result<String> {
    match mode {
        FingerprintMode::None => Ok(String::new()),
        FingerprintMode::Motherboard => get_motherboard_serial(),
        FingerprintMode::CPU => get_cpu_id(),
        FingerprintMode::Combined => {
            let mut components = Vec::new();

            if let Ok(mb) = get_motherboard_serial() {
                components.push(mb);
            }

            if let Ok(cpu) = get_cpu_id() {
                components.push(cpu);
            }

            if components.is_empty() {
                return Err(ShieldError::FingerprintUnavailable);
            }

            // Create hash of combined components (matches SaaSClient)
            let combined = components.join("-");
            Ok(format!("{:x}", md5::compute(combined.as_bytes())))
        }
    }
}

/// Get motherboard serial number (platform-specific).
#[cfg(target_os = "windows")]
fn get_motherboard_serial() -> Result<String> {
    let output = Command::new("wmic")
        .args(&["baseboard", "get", "serialnumber", "/value"])
        .output()
        .map_err(|_| ShieldError::FingerprintUnavailable)?;

    let output_str = String::from_utf8_lossy(&output.stdout);
    for line in output_str.lines() {
        if line.starts_with("SerialNumber=") {
            let serial = line.replace("SerialNumber=", "").trim().to_string();
            if !serial.is_empty() && serial != "To be filled by O.E.M." {
                return Ok(serial);
            }
        }
    }
    Err(ShieldError::FingerprintUnavailable)
}

#[cfg(target_os = "linux")]
fn get_motherboard_serial() -> Result<String> {
    // Try DMI sysfs first (no elevated privileges needed)
    if let Ok(content) = std::fs::read_to_string("/sys/class/dmi/id/board_serial") {
        let serial = content.trim();
        if !serial.is_empty() && serial != "To be filled by O.E.M." {
            return Ok(serial.to_string());
        }
    }

    // Fallback to dmidecode (may require sudo)
    let output = Command::new("dmidecode")
        .args(&["-s", "baseboard-serial-number"])
        .output()
        .map_err(|_| ShieldError::FingerprintUnavailable)?;

    let serial = String::from_utf8_lossy(&output.stdout).trim().to_string();
    if !serial.is_empty() && serial != "To be filled by O.E.M." {
        Ok(serial)
    } else {
        Err(ShieldError::FingerprintUnavailable)
    }
}

#[cfg(target_os = "macos")]
fn get_motherboard_serial() -> Result<String> {
    let output = Command::new("system_profiler")
        .args(&["SPHardwareDataType"])
        .output()
        .map_err(|_| ShieldError::FingerprintUnavailable)?;

    let output_str = String::from_utf8_lossy(&output.stdout);
    for line in output_str.lines() {
        if line.contains("Serial Number") {
            if let Some(serial) = line.split(':').nth(1) {
                return Ok(serial.trim().to_string());
            }
        }
    }
    Err(ShieldError::FingerprintUnavailable)
}

#[cfg(not(any(target_os = "windows", target_os = "linux", target_os = "macos")))]
fn get_motherboard_serial() -> Result<String> {
    Err(ShieldError::FingerprintUnavailable)
}

/// Get CPU identifier (platform-specific).
#[cfg(target_os = "windows")]
fn get_cpu_id() -> Result<String> {
    let output = Command::new("wmic")
        .args(&["cpu", "get", "ProcessorId", "/value"])
        .output()
        .map_err(|_| ShieldError::FingerprintUnavailable)?;

    let output_str = String::from_utf8_lossy(&output.stdout);
    for line in output_str.lines() {
        if line.starts_with("ProcessorId=") {
            let cpu_id = line.replace("ProcessorId=", "").trim().to_string();
            if !cpu_id.is_empty() {
                return Ok(cpu_id);
            }
        }
    }
    Err(ShieldError::FingerprintUnavailable)
}

#[cfg(target_os = "linux")]
fn get_cpu_id() -> Result<String> {
    if let Ok(content) = std::fs::read_to_string("/proc/cpuinfo") {
        // Use first processor line as identifier
        for line in content.lines() {
            if line.starts_with("processor") && line.contains("0") {
                return Ok(format!("{:x}", md5::compute(line.as_bytes())));
            }
        }
    }
    Err(ShieldError::FingerprintUnavailable)
}

#[cfg(target_os = "macos")]
fn get_cpu_id() -> Result<String> {
    let output = Command::new("sysctl")
        .args(&["-n", "machdep.cpu.brand_string"])
        .output()
        .map_err(|_| ShieldError::FingerprintUnavailable)?;

    let cpu_info = String::from_utf8_lossy(&output.stdout).trim().to_string();
    if !cpu_info.is_empty() {
        Ok(format!("{:x}", md5::compute(cpu_info.as_bytes())))
    } else {
        Err(ShieldError::FingerprintUnavailable)
    }
}

#[cfg(not(any(target_os = "windows", target_os = "linux", target_os = "macos")))]
fn get_cpu_id() -> Result<String> {
    Err(ShieldError::FingerprintUnavailable)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_none_mode() {
        let fp = collect_fingerprint(FingerprintMode::None).unwrap();
        assert_eq!(fp, "");
    }

    #[test]
    fn test_combined_mode() {
        // This test may fail on systems without accessible hardware IDs
        // In CI/CD, consider mocking or skipping
        match collect_fingerprint(FingerprintMode::Combined) {
            Ok(fp) => {
                assert!(!fp.is_empty());
                assert_eq!(fp.len(), 32); // MD5 hex string
            }
            Err(ShieldError::FingerprintUnavailable) => {
                // Expected on VMs or restricted environments
            }
            Err(e) => panic!("Unexpected error: {:?}", e),
        }
    }
}

File: java/src/main/java/ai/dikestra/shield/Fingerprint.java
package ai.dikestra.shield;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.security.MessageDigest;
import java.util.ArrayList;
import java.util.List;

/**
 * Hardware fingerprinting for device-bound encryption.
 *
 * <p>Collects platform-specific hardware identifiers to create device-bound keys.
 * Adapted from SaaSClient-SideLicensingSystem.
 *
 * <p>Example:
 * <pre>{@code
 * String fp = Fingerprint.collect(FingerprintMode.COMBINED);
 * Shield shield = Shield.withFingerprint("password", "github.com", FingerprintMode.COMBINED);
 * }</pre>
 */
public class Fingerprint {

    /**
     * Fingerprint collection mode.
     */
    public enum FingerprintMode {
        /** No fingerprinting (backward compatible) */
        NONE,
        /** Motherboard serial only */
        MOTHERBOARD,
        /** CPU identifier only */
        CPU,
        /** Motherboard + CPU (recommended) */
        COMBINED
    }

    /**
     * Collect hardware fingerprint.
     *
     * @param mode Fingerprint mode
     * @return Fingerprint string (SHA-256 hex), or empty for NONE
     * @throws Exception If hardware identifiers cannot be collected
     */
    public static String collect(FingerprintMode mode) throws Exception {
        if (mode == FingerprintMode.NONE) {
            return "";
        }

        if (mode == FingerprintMode.MOTHERBOARD) {
            return getMotherboardSerial();
        }

        if (mode == FingerprintMode.CPU) {
            return getCpuId();
        }

        if (mode == FingerprintMode.COMBINED) {
            List<String> components = new ArrayList<>();

            try {
                components.add(getMotherboardSerial());
            } catch (Exception e) {
                // Continue without motherboard
            }

            try {
                components.add(getCpuId());
            } catch (Exception e) {
                // Continue without CPU
            }

            if (components.isEmpty()) {
                throw new Exception("Hardware fingerprint unavailable");
            }

            // Create SHA-256 hash of combined components
            String combined = String.join("-", components);
            return sha256Hex(combined);
        }

        throw new IllegalArgumentException("Unknown fingerprint mode: " + mode);
    }

    /**
     * Get motherboard serial number (platform-specific).
     */
    private static String getMotherboardSerial() throws Exception {
        String os = System.getProperty("os.name").toLowerCase();

        if (os.contains("win")) {
            Process process = Runtime.getRuntime().exec(
                new String[]{"wmic", "baseboard", "get", "serialnumber", "/value"}
            );

            try (BufferedReader reader = new BufferedReader(
                    new InputStreamReader(process.getInputStream()))) {
                String line;
                while ((line = reader.readLine()) != null) {
                    if (line.startsWith("SerialNumber=")) {
                        String serial = line.replace("SerialNumber=", "").trim();
                        if (!serial.isEmpty() && !serial.equals("To be filled by O.E.M.")) {
                            return serial;
                        }
                    }
                }
            }
        } else if (os.contains("linux")) {
            // Try DMI sysfs first
            try {
                String serial = new String(Files.readAllBytes(
                    Paths.get("/sys/class/dmi/id/board_serial")
                )).trim();
                if (!serial.isEmpty() && !serial.equals("To be filled by O.E.M.")) {
                    return serial;
                }
            } catch (Exception e) {
                // Fall through to dmidecode
            }

            // Fallback to dmidecode
            Process process = Runtime.getRuntime().exec(
                new String[]{"dmidecode", "-s", "baseboard-serial-number"}
            );

            try (BufferedReader reader = new BufferedReader(
                    new InputStreamReader(process.getInputStream()))) {
                String serial = reader.readLine();
                if (serial != null) {
                    serial = serial.trim();
                    if (!serial.isEmpty() && !serial.equals("To be filled by O.E.M.")) {
                        return serial;
                    }
                }
            }
        } else if (os.contains("mac")) {
            Process process = Runtime.getRuntime().exec(
                new String[]{"system_profiler", "SPHardwareDataType"}
            );

            try (BufferedReader reader = new BufferedReader(
                    new InputStreamReader(process.getInputStream()))) {
                String line;
                while ((line = reader.readLine()) != null) {
                    if (line.contains("Serial Number")) {
                        String[] parts = line.split(":");
                        if (parts.length >= 2) {
                            return parts[1].trim();
                        }
                    }
                }
            }
        }

        throw new Exception("Motherboard serial number unavailable");
    }

    /**
     * Get CPU identifier (platform-specific).
     */
    private static String getCpuId() throws Exception {
        String os = System.getProperty("os.name").toLowerCase();

        if (os.contains("win")) {
            Process process = Runtime.getRuntime().exec(
                new String[]{"wmic", "cpu", "get", "ProcessorId", "/value"}
            );

            try (BufferedReader reader = new BufferedReader(
                    new InputStreamReader(process.getInputStream()))) {
                String line;
                while ((line = reader.readLine()) != null) {
                    if (line.startsWith("ProcessorId=")) {
                        String cpuId = line.replace("ProcessorId=", "").trim();
                        if (!cpuId.isEmpty()) {
                            return cpuId;
                        }
                    }
                }
            }
        } else if (os.contains("linux")) {
            String content = new String(Files.readAllBytes(Paths.get("/proc/cpuinfo")));
            for (String line : content.split("\n")) {
                if (line.startsWith("processor") && line.contains("0")) {
                    // Use first processor as identifier (hashed)
                    return sha256Hex(line);
                }
            }
        } else if (os.contains("mac")) {
            Process process = Runtime.getRuntime().exec(
                new String[]{"sysctl", "-n", "machdep.cpu.brand_string"}
            );

            try (BufferedReader reader = new BufferedReader(
                    new InputStreamReader(process.getInputStream()))) {
                String cpuInfo = reader.readLine();
                if (cpuInfo != null) {
                    cpuInfo = cpuInfo.trim();
                    if (!cpuInfo.isEmpty()) {
                        return sha256Hex(cpuInfo);
                    }
                }
            }
        }

        throw new Exception("CPU identifier unavailable");
    }

    /**
     * SHA-256 hash helper.
     */
    private static String sha256Hex(String input) throws Exception {
        MessageDigest md = MessageDigest.getInstance("SHA-256");
        byte[] digest = md.digest(input.getBytes());

        StringBuilder hex = new StringBuilder();
        for (byte b : digest) {
            hex.append(String.format("%02x", b));
        }
        return hex.toString();
    }
}
